backend/compound_library/src/resolvers/compound_types_res.rs

```diff
@@ -0,0 +1,77 @@
+// src/resolvers/compound_types_res.rs
+
+use crate::entities::{compound_instances, compound_types};
+use async_graphql::{ComplexObject, Context, Object};
+use opa_client::subject_authorization;
+use sea_orm::{ActiveValue, DatabaseConnection, EntityTrait, ModelTrait};
+use the_paginator::graphql::{CursorInput, ModelConnection};
+use uuid::Uuid;
+
+#[derive(Debug, Clone, Default)]
+pub struct CompoundQuery;
+
+#[derive(Debug, Clone, Default)]
+pub struct CompoundMutation;
+
+#[ComplexObject]
+impl compound_types::Model {
+    async fn compound_instances(
+        &self,
+        ctx: &Context<'_>,
+    ) -> async_graphql::Result<Vec<compound_instances::Model>> {
+        subject_authorization!("xchemlab.compound_library.read_compound", ctx).await?;
+        let db = ctx.data::<DatabaseConnection>()?;
+        Ok(self
+            .find_related(compound_instances::Entity)
+            .all(db)
+            .await?)
+    }
+}
+
+#[Object]
+impl CompoundQuery {
+    async fn compounds(
+        &self,
+        ctx: &Context<'_>,
+        cursor: CursorInput,
+    ) -> async_graphql::Result<ModelConnection<compound_types::Model>> {
+        subject_authorization!("xchemlab.compound_library.read_compound", ctx).await?;
+        let db = ctx.data::<DatabaseConnection>()?;
+        Ok(cursor
+            .try_into_query_cursor::<compound_types::Entity>()?
+            .all(db)
+            .await?
+            .try_into_connection()?)
+    }
+
+    async fn compound(
+        &self,
+        ctx: &Context<'_>,
+        id: Uuid,
+    ) -> async_graphql::Result<Option<compound_types::Model>> {
+        subject_authorization!("xchemlab.compound_library.read_compound", ctx).await?;
+        let db = ctx.data::<DatabaseConnection>()?;
+        Ok(compound_types::Entity::find_by_id(id).one(db).await?)
+    }
+}
+
+#[Object]
+impl CompoundMutation {
+    async fn add_compound(
+        &self,
+        ctx: &Context<'_>,
+        name: String,
+        smiles: String,
+    ) -> async_graphql::Result<compound_types::Model> {
```

Comment:
Do we want to record who added this and when?

Reply:
Record in the database?

Reply:
Yeah